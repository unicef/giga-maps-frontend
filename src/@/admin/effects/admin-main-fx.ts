import { createEffect } from "effector"

import { createRequestAuthFx } from "~/core/auth/effects/common.fx"

import { ApiConfig, InvalidateCache } from "../types/giga-layer.type"

export const getAppConfigValuesFx = createEffect(() => {
  return createRequestAuthFx({
    url: `accounts/app_configs/`
  }) as Promise<ApiConfig>
})

export const getInvalidateCacheFx = createEffect((data: object) => {
  return createRequestAuthFx({
    url: `accounts/invalidate-cache-patterns/`,
    method: 'DELETE',
    data
  }) as Promise<InvalidateCache>
})