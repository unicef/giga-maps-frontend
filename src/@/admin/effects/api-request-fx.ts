import { createEffect } from "effector";

import { APIListType } from "~/api/types";
import { createRequestAuthFx } from "~/core/auth/effects/common.fx";

import { ApiKeysAdminRequestType } from "../types/api-request.type";

export const getCountryApiFx = createEffect((query: string) => {
  return createRequestAuthFx({
    url: `locations/countries/${query}`,
  }) as Promise<ApiKeysAdminRequestType[]>
})

export const getAllApiKeyRequest = createEffect(({ page, pageSize, countryIds, search }: { page: number; pageSize: number, countryIds?: string, search?: string }) => {
  return createRequestAuthFx({
    url: `accounts/api_keys/?page_size=${pageSize}&page=${page}${countryIds ? `&country_id__in=${countryIds}` : ''}${search ? `&search=${search}` : ''}&expand=user,api&ordering=-status`
  }) as Promise<APIListType<ApiKeysAdminRequestType>>
})

export const changeApiKeyRequestFx = createEffect(({ body, id }: { body: any; id: number }) => {
  return createRequestAuthFx({
    url: `accounts/api_keys/${id}/`,
    data: body,
    method: 'PUT'
  }) as Promise<ApiKeysAdminRequestType>
})

export const deleteApiKeyRequestFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `accounts/api_keys/${id}/`,
    method: 'DELETE'
  }) as Promise<ApiKeysAdminRequestType>
})

export const updateApiKeyExtensionFx = createEffect(({ id, body }: { id: number, body: any; }) => {
  return createRequestAuthFx({
    url: `accounts/api_keys/${id}/`,
    method: 'PUT',
    data: body
  }) as Promise<ApiKeysAdminRequestType>
});