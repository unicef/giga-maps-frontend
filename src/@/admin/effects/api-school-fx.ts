import { createEffect } from "effector"

import { APIListType } from "~/api/types"
import { createRequestAuthFx } from "~/core/auth/effects/common.fx"

import { AdminType, CsvImport, SchoolDailyType, SchoolSummaryType, SchoolType } from "../types/school.type";



export const getSchoolListFx = createEffect(({ page, pageSize, search, filter }: { page?: number; pageSize?: number; search?: string, filter?: number[] }) => {
  let url = `locations/schools/school/?`
  url += `page=${page}&page_size=${pageSize}`;

  if (filter && filter.length > 0) {
    url += `&country_id__in=${filter.join(',')}`;
  }

  if (search) {
    url += `&search=${search}`;
  }

  return createRequestAuthFx({
    url: url
  }) as Promise<APIListType<SchoolType>>
})

export const createOrUpdateSchoolFx = createEffect(({ body, isEditMode, params }: any) => {
  return createRequestAuthFx({
    method: isEditMode ? 'PUT' : 'POST',
    url: `locations/schools/school/${params.id ? params.id + '/' : ''}`,
    data: body
  }) as Promise<SchoolType>
})

export const getSchoolIdFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `locations/schools/school/${id + '/'}`
  }) as Promise<SchoolType>
})

export const deleteSchoolFx = createEffect(({ body, }: any) => {
  return createRequestAuthFx({
    method: 'DELETE',
    url: `locations/schools/school/`,
    data: body
  }) as Promise<SchoolType>
})


export const getSchoolSummaryListFx = createEffect(({ page, pageSize, search, filter, schoolId }: { page?: number; pageSize?: number; search?: string, filter?: number[], schoolId?: number }) => {
  return createRequestAuthFx({
    url: `/statistics/schoolweeklystatus/?page=${page}&page_size=${pageSize}${(filter && filter.length > 0) ? `&country_id=${filter}` : ''}${search ? `&search=${search}` : ''}${schoolId ? `&school_id=${schoolId}` : ''}`
  }) as Promise<APIListType<SchoolSummaryType>>
})

export const getSchoolSummaryIdFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `/statistics/schoolweeklystatus/${id + '/'}`
  }) as Promise<SchoolSummaryType>
})

export const deleteSchoolSummaryFx = createEffect(({ body, }: any) => {
  return createRequestAuthFx({
    method: 'DELETE',
    url: `/statistics/schoolweeklystatus/`,
    data: body
  }) as Promise<SchoolType>
})

export const createOrUpdateSchoolSummaryFx = createEffect(({ body, isEditMode, params }: any) => {
  return createRequestAuthFx({
    method: isEditMode ? 'PUT' : 'POST',
    url: `/statistics/schoolweeklystatus/${params.id ? params.id + '/' : ''}`,
    data: body
  }) as Promise<SchoolType>
})

export const getSchoolDailyListFx = createEffect(({ page, pageSize, search, filter }: { page?: number; pageSize?: number; search?: string, filter?: number[] }) => {
  return createRequestAuthFx({
    url: `/statistics/schooldailystatus/?page=${page}&page_size=${pageSize}${(filter && filter.length > 0) ? `&country_id=${filter}` : ''}${search ? `&search=${search}` : ''}`
  }) as Promise<APIListType<SchoolDailyType>>
})

export const getSchoolDailyIdFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `/statistics/schooldailystatus/${id + '/'}`
  }) as Promise<SchoolSummaryType>
})

export const deleteSchoolDailyFx = createEffect(({ body, }: any) => {
  return createRequestAuthFx({
    method: 'DELETE',
    url: `statistics/schooldailystatus/`,
    data: body
  }) as Promise<SchoolSummaryType>
})

export const createOrUpdateSchoolDailyFx = createEffect(({ body, isEditMode, params }: any) => {
  return createRequestAuthFx({
    method: isEditMode ? 'PUT' : 'POST',
    url: `statistics/schooldailystatus/${params.id ? params.id + '/' : ''}`,
    data: body
  }) as Promise<SchoolType>
})


export const importCsvFx = createEffect(({ formData }: any) => {
  return createRequestAuthFx({
    method: 'POST',
    url: `/locations/schools/fileimport/`,
    body: formData
  }) as Promise<CsvImport>
})

export const getCsvImportListFx = createEffect(({ page, pageSize, search }: { page?: number; pageSize?: number; search?: string }) => {
  return createRequestAuthFx({
    url: `locations/schools/fileimport/?page=${page}&page_size=${pageSize}${search ? `&search=${search}` : ''}`
  }) as Promise<APIListType<CsvImport>>
})

export const getAdmin1ListFx = createEffect(({ layerName, countryId }: { layerName?: string; countryId?: string; }) => {
  return createRequestAuthFx({
    url: `locations/country-admin-metadata/?layer_name=${layerName}&country_id=${countryId}&ordering=description,name&expand=country,parent&page_size=500&fields=id,name,description,name_en&page=1`
  }) as Promise<APIListType<AdminType>>
})


export const getAdmin2ListFx = createEffect(({ layerName, countryId, parentId }: { layerName?: string; countryId?: string; parentId?: number }) => {
  return createRequestAuthFx({
    url: `locations/country-admin-metadata/?layer_name=${layerName}&country_id=${countryId}&ordering=description,name&expand=country,parent${parentId ? `&parent_id=${parentId}` : ''}&page_size=500&fields=id,name,description,name_en&page=1`
  }) as Promise<APIListType<AdminType>>
})