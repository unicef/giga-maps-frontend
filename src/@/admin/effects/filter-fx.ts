import { createEffect } from "effector"
import { createRequestAuthFx } from "~/core/auth/effects/common.fx"
import { APIListType } from "~/api/types"
import { FilterConfiguration, FilterListType } from "../types/filter-list.type"
import { FilterAllValueType } from "../types/filter-list-type"

export const getFilterListFx = createEffect(({ page, pageSize, search }: { page: number, pageSize: number, search?: string }) => {
  const query = new URLSearchParams()
  query.set('page_size', String(pageSize))
  query.set('page', String(page))
  if (search) {
    query.set('search', search)
  }
  return createRequestAuthFx({
    url: `accounts/adv_filters/?${query.toString()}&ordering=-last_modified_at&expand=column_configuration,published_by`
  }) as Promise<APIListType<FilterListType>>
})

export const getFilterPublishedListFx = createEffect(() => {
  return createRequestAuthFx({
    url: `accounts/adv_filters/?status=PUBLISHED`
  }) as Promise<APIListType<FilterListType>>
})

export const filterColumnListFx = createEffect(() => {
  return createRequestAuthFx({
    url: `accounts/column_configurations/?is_filter_applicable=true`
  }) as Promise<APIListType<FilterConfiguration>>
})

export const deleteFilterFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `accounts/adv_filters/${id}/`,
    method: 'DELETE'
  }) as Promise<FilterListType>
})

export const addFilterFx = createEffect(({ body }: { body: FilterAllValueType }) => {
  return createRequestAuthFx({
    url: `accounts/adv_filters/`,
    method: 'POST',
    data: body
  }) as Promise<FilterListType>
})
export const editFilterFx = createEffect(({ id, body }: { id: number, body: FilterAllValueType | { status: string } }) => {
  return createRequestAuthFx({
    url: `accounts/adv_filters/${id + '/'}`,
    method: 'PUT',
    data: body
  }) as Promise<FilterListType>
})

export const getFilterListIdFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `accounts/adv_filters/?id=${id}`
  }) as Promise<APIListType<FilterListType>>
})

export const publishFilterFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    method: 'PUT',
    url: `accounts/adv_filters/${id}/publish/`,
  }) as Promise<FilterListType>
})