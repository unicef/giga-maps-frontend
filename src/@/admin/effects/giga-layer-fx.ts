import { attach, createEffect } from "effector"

import { APIListType } from "~/api/types"
import { createRequestAuthFx } from "~/core/auth/effects/common.fx"

import { DataLayer, DataSource, LayerStatusType, PreviewDataType } from "../types/giga-layer.type"

export const commonLayerListFx = createEffect(({ page, pageSize, search }: { page: number, pageSize: number; search?: string }) => {
  return createRequestAuthFx({
    url: `accounts/layers/?expand=created_by,last_modified_by,published_by&page_size=${pageSize}&page=${page}${search ? `&search=${search}` : ''}&ordering=-last_modified_at,name`
  }) as Promise<APIListType<DataLayer>>
})

export const cacheDataLayerListFx = attach({
  effect: commonLayerListFx,
  mapParams: (props) => props,
})

export const getDataLayerListFx = attach({
  effect: commonLayerListFx,
  mapParams: (props) => props,
})


export const getDataLayerByIdFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `accounts/layers/?id=${id}&expand=created_by,last_modified_by,published_by`
  }) as Promise<APIListType<DataLayer>>
})

export const createDataLayerFx = createEffect(({ body, isEditMode, params }: any) => {
  return createRequestAuthFx({
    method: isEditMode ? 'PUT' : 'POST',
    url: `accounts/layers/${params.id ? params.id + '/' : ''}`,
    data: body
  }) as Promise<DataLayer>
})

export const getApiSourceValuesFx = createEffect(({ type }: { type: string }) => {
  return createRequestAuthFx({
    url: `accounts/data_sources/?data_source_type__in=${type}&status=PUBLISHED`
  }) as Promise<APIListType<DataSource>>
})

export const publishDataLayerFx = createEffect(({ id, status }: { id: number, status: LayerStatusType }) => {
  return createRequestAuthFx({
    method: 'PUT',
    data: { status },
    url: `accounts/layers/${id}/publish/`,
  }) as Promise<DataLayer>
})

export const getDataPreviewFx = createEffect(({ id }: { id: number }) => {
  return createRequestAuthFx({
    url: `/accounts/layers/${id}/preview/`,
  }) as Promise<PreviewDataType>
})
