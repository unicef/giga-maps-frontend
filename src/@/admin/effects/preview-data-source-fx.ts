import { createEffect } from "effector";
import mapboxgl, { Map } from "mapbox-gl";

import { PreviewDataType } from "../types/giga-layer.type";
import { addCountryBoundries } from "./map-country-boundary.fx";
import { DEFAULT_SOURCE } from "~/@/map/map.constant";

const getSchoolsGeoJson = (data: PreviewDataType['map']) => {
  return {
    type: 'FeatureCollection',
    features: data.map((item: PreviewDataType['map'][0]) => ({
      type: 'Feature',
      properties: {
        description:
          `<strong>${item.school_name}</strong>
        <p>Giga id: ${item.school_id_giga}</p>
        <p>Lat: ${item.latitude}, Lon: ${item.longitude}</p>`
      },
      geometry: {
        'type': 'Point',
        'coordinates': [item.longitude, item.latitude]
      },
      id: item.id,
    })),
  };
};


export const plotMapData = (map: Map, mapData: PreviewDataType['map']) => {
  const data = getSchoolsGeoJson(mapData) as unknown as GeoJSON.Geometry;
  map.addSource(DEFAULT_SOURCE, {
    type: 'geojson',
    data
  });

  map.addLayer({
    'id': 'places',
    'type': 'circle',
    'source': DEFAULT_SOURCE,
    'paint': {
      'circle-color': '#4264fb',
      'circle-radius': 6,
      'circle-stroke-width': 2,
      'circle-stroke-color': '#ffffff'
    }
  });

  const popup = new mapboxgl.Popup({
    closeButton: true,
    closeOnClick: true,
    className: 'preview-data-popup'
  });

  const createPopup = (e: any) => {
    const coordinates = e.features[0].geometry.coordinates.slice();
    const description = e.features[0].properties.description;

    // Ensure that if the map is zoomed out such that multiple
    // copies of the feature are visible, the popup appears
    // over the copy being pointed to.
    while (Math.abs(e.lngLat.lng - coordinates[0]) > 180) {
      coordinates[0] += e.lngLat.lng > coordinates[0] ? 360 : -360;
    }

    // Populate the popup and set its coordinates
    // based on the feature found.
    popup.setLngLat(coordinates).setHTML(`<div class="map-content-popup">${description}</div>`).addTo(map);
  }

  map.on('mouseenter', 'places', (e) => {
    // Change the cursor style as a UI indicator.
    map.getCanvas().style.cursor = 'pointer';
    // Copy coordinates array.
    createPopup(e);
  })
    .on('click', 'places', (e) => createPopup(e))

  map.on('mouseleave', 'places', () => { });
}

export const previewDataSourceFx = createEffect(({ map, mapData }: { map: Map, mapData: PreviewDataType['map'] }) => {
  if (!map) return;
  try {
    addCountryBoundries(map);
    plotMapData(map, mapData);
  } catch (e) {
    console.log(e);
  }
})
