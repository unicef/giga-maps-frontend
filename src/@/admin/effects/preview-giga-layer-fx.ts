import { createEffect } from "effector";
import { Map } from "mapbox-gl";

import { DEFAULT_SOURCE, LayerDataProps, stylePaintData } from "~/@/map/map.constant";
import { animateCircles, checkSourceAvailable, createSchoolLayer, createSchoolSource, createSelectedLayer, deleteSourceAndLayers } from "~/@/map/utils";
import { Layers, SCHOOL_STATUS_LAYER } from "~/@/sidebar/sidebar.constant";

import { DataLayer, PreviewDataType } from "../types/giga-layer.type";
import { addCountryBoundries } from "./map-country-boundary.fx";
import { isLiveLayer } from "~/@/sidebar/sidebar.util";
import { changeAdminMap } from "~/@/common/admin-map-preview/admin-map.model";

let adminAnimateCircleHandler = { requestId: 0 }; // to clear animation;

const getSchoolsGeoJson = (data: PreviewDataType['map']) => {
  return {
    type: 'FeatureCollection',
    features: data.map((item: PreviewDataType['map'][0]) => ({
      type: 'Feature',

      properties: {
        id: item.name,
        [LayerDataProps.connectivityStatus.key]: item.connectivity_status,
        [LayerDataProps.connectivity.key]: item.connectivity,
        [LayerDataProps.coverage.key]: item?.field_status,
        is_rt_connected: item?.is_rt_connected
      },
      geometry: item.geopoint,
      id: item.id,
    })),
  };
}

export const plotMapData = ({ map, mapData, currentLayerItem }: { map: Map; mapData: PreviewDataType['map'], currentLayerItem: DataLayer }) => {
  if (checkSourceAvailable(map, DEFAULT_SOURCE)) return null;
  createSchoolSource({ map, schoolData: getSchoolsGeoJson(mapData) });
  const isLive = !!isLiveLayer(currentLayerItem.type);

  createSelectedLayer(map, {
    id: Layers.connectivity,
    isLive,
    isDynamicLayer: false,
    paintData: stylePaintData.dark,
    mapRoute: { map: true, country: false, schools: false },
    options: {}
  });
  if (isLive) {
    adminAnimateCircleHandler = animateCircles({ map, id: Layers.connectivity });
  }
  // create school layer;
  createSchoolLayer(map, {
    id: String(SCHOOL_STATUS_LAYER.id),
    paintData: stylePaintData.dark,
    mapRoute: { map: true, country: false, schools: false },
    options: {}
  });

}

export const previewDataLayerFx = createEffect(({ map, mapData, currentLayerItem }: { map: Map; mapData: PreviewDataType['map'], currentLayerItem: DataLayer }) => {
  if (!map) return;
  try {
    addCountryBoundries(map);
    plotMapData({ map, mapData, currentLayerItem });
  } catch (e) {
    console.log(e);
  }
})

export const clearAdminMapData = createEffect(({ map }: { map: Map }) => {
  clearTimeout(adminAnimateCircleHandler.requestId);
  deleteSourceAndLayers({ map, sourceId: DEFAULT_SOURCE });
  changeAdminMap(null);
});