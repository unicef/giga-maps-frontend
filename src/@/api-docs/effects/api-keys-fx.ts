import { createEffect } from "effector";

import { APIListType } from "~/api/types";
import { createRequestAuthFx } from "~/core/auth/effects/common.fx";

import { ApiKeysType } from "../types/api-keys.type";
import { ExploreApiType } from "../types/explore-api-type";

export const requestForApiKeyFx = createEffect((body: any) => {
  return createRequestAuthFx({
    url: 'accounts/api_keys/',
    method: 'POST',
    data: body,
  }) as Promise<APIListType<ExploreApiType>>
});

export const getApiKeyListFx = createEffect(({ userId, page, pageSize }: { userId?: number, page?: number; pageSize?: number; }) => {
  return createRequestAuthFx({
    url: `accounts/api_keys/?page=${page}&page_size=${pageSize}&expand=api&ordering=-last_modified_at&user_id=${userId}`,
    method: 'GET',
  }) as Promise<APIListType<ApiKeysType>>
});

export const validateApiKeyFx = createEffect((body: any) => {
  return createRequestAuthFx({
    url: `accounts/validate_api_key/`,
    method: 'PUT',
    data: body
  }) as Promise<ApiKeysType['active_api_categories_list']>
});

export const requestForExtensionFx = createEffect(({ id, body }: { id?: number, body: any; }) => {
  return createRequestAuthFx({
    url: `/accounts/api_keys/${id}/request_extension/`,
    method: 'PUT',
    data: body
  }) as Promise<ApiKeysType>
});
