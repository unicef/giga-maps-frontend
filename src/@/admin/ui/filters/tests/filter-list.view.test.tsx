import { fireEvent, getByRole, render, waitFor } from "@testing-library/react";
import { testWrapper } from "~/tests/jest-wrapper";
import ListFilterView from "../filter-list.view";
import AddFilterList from "../add-filter.view";
import AdminFilters from "..";
import { adminFilterRoute } from "~/core/routes";
import { deleteFilterFx, getFilterListFx, publishFilterFx } from "~/@/admin/effects/filter-fx";
import { before } from "node:test";
import { fetchMockResponse } from "~/tests/fetchMock";

describe('Admin Filter', () => {

  beforeEach(() => {
    fetchMock.mockResponse(fetchMockResponse)
  })
  it('should render component', () => {
    const { asFragment } = render(testWrapper(<ListFilterView />));
    expect(asFragment).toMatchSnapshot();
  });

  it('should render component', () => {
    const { asFragment } = render(testWrapper(<AddFilterList />));
    expect(asFragment).toMatchSnapshot();
  });

  it('should render component', () => {
    const { asFragment } = render(testWrapper(<AdminFilters />));
    expect(asFragment).toMatchSnapshot();
  });

  it('should handle search value changes', async () => {
    adminFilterRoute.navigate();
    const { getByPlaceholderText } = render(testWrapper(<ListFilterView />));
    const searchInput = getByPlaceholderText('Filter by name, type');
    fireEvent.change(searchInput, { target: { value: 'test search' } });
    expect(searchInput.value).toBe('test search');
  });

});