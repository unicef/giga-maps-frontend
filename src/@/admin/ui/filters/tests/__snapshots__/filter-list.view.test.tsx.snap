// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Filter should render component 1`] = `[Function]`;

exports[`Admin Filter should render component 2`] = `[Function]`;

exports[`Admin Filter should render component 3`] = `[Function]`;
